### PR TITLE
fix bug when output tensor is 0 shape

### PR DESCRIPTION
**Pull request by liufengwei0103**
1.判断在output tensor element cnt为0时返回。
2.增加测例两个，分别是被指定concat维度是0和output本身为0。

### CODE REVIEW

File: oneflow/user/kernels/concat_kernel.cpp [hunk 1]
```diff
@@ -51,8 +51,10 @@ class ConcatKernel final : public user_op::OpKernel {
 
   void Compute(user_op::KernelComputeContext* ctx) const override {
     user_op::Tensor* out_tensor = ctx->Tensor4ArgNameAndIndex("out", 0);
+    if (out_tensor->shape().elem_cnt() == 0) { return; }
     const int64_t axis = ctx->Attr<int64_t>("axis");
     const int64_t out_cols = out_tensor->shape().Count(axis);
+    CHECK_GT(out_cols, 0);
```
Comment (wyg1997) [inline oneflow/user/kernels/concat_kernel.cpp:57]: 这个不用了，如果这个是 0 的话，上面那个必定会 return

Comment (liufengwei0103) [reply oneflow/user/kernels/concat_kernel.cpp]: > 这个不用了，如果这个是 0 的话，上面那个必定会 return

删了

File: python/oneflow/test/modules/test_concat.py [hunk 2]
```diff
@@ -135,13 +135,28 @@ def test_concat(test_case):
...
+        y = random_pytorch_tensor(4, 2, 0, 2, 4).to(device)
+        z = torch.cat((x, y), dim=2)
+        return z
+
+    @autotest(n=10, auto_backward=False)
+    def test_concat_with_0shape_specified_axis(test_case):
+        device = random_device()
+        x = random_pytorch_tensor(4, 2, 0, 2, 4).to(device)
+        y = random_pytorch_tensor(4, 2, 0, 2, 4).to(device)
+        z = torch.cat((x, y), dim=1)
```
Comment (wyg1997) [inline python/oneflow/test/modules/test_concat.py:158]: 这个 dim 改成 random(0, 4) 怎么样，这样就什么情况都测到了

Comment (liufengwei0103) [reply python/oneflow/test/modules/test_concat.py]: > 这个 dim 改成 random(0, 4) 怎么样，这样就什么情况都测到了

已删

**Review by wyg1997 — APPROVED** [review]

**Comment by liufengwei0103** [comment]
test_indexing 测例，compare_getitem_with_numpy(x, se[-1:]) 方法，暂时看不出逻辑错误， 排除测例之间的相互影响，本地不能复现，再在ci上跑一次。

**Comment by wyg1997** [comment]
> test_indexing 测例，compare_getitem_with_numpy(x, se[-1:]) 方法，暂时看不出逻辑错误， 排除测例之间的相互影响，本地不能复现，再在ci上跑一次。

之前有遇到过这个单测有概率挂的情况

**Comment by hjchen2** [comment]
> > test_indexing 测例，compare_getitem_with_numpy(x, se[-1:]) 方法，暂时看不出逻辑错误， 排除测例之间的相互影响，本地不能复现，再在ci上跑一次。
> 
> 之前有遇到过这个单测有概率挂的情况

这个问题有可能是精度导致的，因为

> > test_indexing 测例，compare_getitem_with_numpy(x, se[-1:]) 方法，暂时看不出逻辑错误， 排除测例之间的相互影响，本地不能复现，再在ci上跑一次。
> 
> 之前有遇到过这个单测有概率挂的情况

这个单测我觉得最好还是输入用rand接口生成，现在直接是empty tensor，就可能出现nan != nan的情况。

**Comment by wyg1997** [comment]
> > > test_indexing 测例，compare_getitem_with_numpy(x, se[-1:]) 方法，暂时看不出逻辑错误， 排除测例之间的相互影响，本地不能复现，再在ci上跑一次。
> > 
> > 
> > 之前有遇到过这个单测有概率挂的情况
> 
> 这个问题有可能是精度导致的，因为
> 
> > > test_indexing 测例，compare_getitem_with_numpy(x, se[-1:]) 方法，暂时看不出逻辑错误， 排除测例之间的相互影响，本地不能复现，再在ci上跑一次。
> > 
> > 
> > 之前有遇到过这个单测有概率挂的情况
> 
> 这个单测我觉得最好还是输入用rand接口生成，现在直接是empty tensor，就可能出现nan != nan的情况。

嗯嗯，我和丰伟已经讨论着用其它接口了，empty 就是一片未初始化的内存，这里用不合适

**Comment by hjchen2** [comment]
> 嗯嗯，我和丰伟已经讨论着用其它接口了，empty 就是一片未初始化的内存，这里用不合适

好的。
